app-backend/api/src/main/scala/datasource/Routes.scala

```diff
@@ -162,19 +162,23 @@ trait DatasourceRoutes extends Authentication
     }
 
   def listUserDatasourceActions(datasourceId: UUID): Route = authenticate { user =>
-    onSuccess(
-      DatasourceDao.getDatasourceById(datasourceId, user).transact(xa).unsafeToFuture
-    ) { datasourceO =>
-      datasourceO match {
-        case Some(datasource) =>
-          if (user.isSuperuser || datasource.owner == user.id) {
-            complete(List("*"))
-          } else {
-            complete {
-              AccessControlRuleDao.listUserActions(user, ObjectType.Datasource, datasourceId).transact(xa).unsafeToFuture
+    authorizeAsync {
+      DatasourceDao.query.authorized(user, ObjectType.Datasource, datasourceId, ActionType.View)
+        .transact(xa).unsafeToFuture
+    } { onSuccess(
+        DatasourceDao.getDatasourceById(datasourceId, user).transact(xa).unsafeToFuture
```

Comment:
At this point, you can assume that the datasource exists, since they have permission to view it. @jisantuc is this a situation where unsafeGetById is warranted?